python/pyarrow/_io.pyx

```diff
@@ -936,6 +936,44 @@ cdef class HdfsFile(NativeFile):
 # ----------------------------------------------------------------------
 # File and stream readers and writers
 
+class _StreamSinkWrapper(object):
+    """
+    Wrapper for writable stream to ensure implements "tell". This will
+    initialize stream at position 0 and increment after each write.
+    """
+
+    def __init__(self, sink):
+        self._sink = sink
+        self._position = 0
+
+    def close(self):
+        self._sink.close()
+
+    def flush(self):
+        self._sink.flush()
+
+    def tell(self):
+        return self._position
+
+    def write(self, data):
+        data = str(data)
+        self._sink.write(data)
+        self._position += len(data)
+
+    @staticmethod
+    def check(sink):
+        import socket
+        if isinstance(sink, socket.socket):
+            # Use a buffer of 0 to flush after each write, otherwise it is
+            # possible to close socket without flushing
+            sink_file = sink.makefile("wb", 0)
+            return _StreamSinkWrapper(sink_file)
+        if not hasattr(sink, "tell"):
+            return _StreamSinkWrapper(sink)
+        else:
+            return sink
```

Comment:
Instead of using this wrapper class, can you instead add `int64_t position_` member to `PyOutputStream` here https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/io.cc#L192, initalized to 0, and return that `position_` from `Tell` instead of calling the file's tell method? Then in `PyOutputStream::Write`, you would increment the position.